### ISSUE TITLE
remote docker connections (tcp) currently not working

### ISSUE THREAD

**Issue by rmoriz** [someara/kitchen-dokken#89]
Just to make it clear that the current kitchen-dokken version does not work if one decides to use a remote dockerd, e.g. using tcp instead of a unix socket.

Also related: https://github.com/someara/kitchen-dokken/issues/79

It looks like there are local volume mounts which by design don't work remotely.

**Comment by someara** [comment]
bizarre... it's explicitly tested against tcp here

https://github.com/someara/kitchen-dokken/blob/master/test/cookbooks/dokken_test/recipes/default.rb#L64

When using tcp, it uses the old 1.x style "data container" with ssh/rsync
https://github.com/someara/kitchen-dokken/blob/master/lib/kitchen/driver/dokken.rb#L69-L72

and

https://github.com/someara/kitchen-dokken/blob/master/lib/kitchen/helpers.rb#L41

Can you help me recreate your problem?

-s

**Comment by rmoriz** [comment]
https://github.com/someara/kitchen-dokken/issues/44 has some details

**Comment by rmoriz** [comment]
Maybe existence of a local sandbox due to a prior, local docker run, might intrigue dokken to assume a local volume mount would be possible. 

I use docker for mac ("local" docker) for small things and a powerful dual xeon box for heavy "remote" docker work. I'll clean the '~/.dokken' and try again next week.

**Comment by rmoriz** [comment]
nope.

```shell
➜  rabbitmq-chef-cookbook git:(rmoriz/inspec) rm -rf ~/.dokken
➜  rabbitmq-chef-cookbook git:(rmoriz/inspec) rm -rf ~/.kitchen/*
➜  rabbitmq-chef-cookbook git:(rmoriz/inspec) eval $(docker-machine env rambo)
➜  rabbitmq-chef-cookbook git:(rmoriz/inspec) env | grep -i docker
DOCKER_TLS_VERIFY=1
DOCKER_HOST=tcp://10.0.1.49:2376
DOCKER_CERT_PATH=/Users/rmoriz/.docker/machine/machines/rambo
DOCKER_MACHINE_NAME=rambo
➜  rabbitmq-chef-cookbook git:(rmoriz/inspec) kitchen setup lwrps-centos-67
-----> Starting Kitchen (v1.16.0)
WARN: Unresolved specs during Gem::Specification.reset:
      artifactory (>= 0)
      winrm (~> 2.0)
      logging (< 3.0, >= 1.6.1)
      rainbow (~> 2)
      diff-lcs (< 2.0, >= 1.2.0)
      unf_ext (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
-----> Creating <lwrps-centos-67>...
       Creating kitchen sandbox at /Users/rmoriz/.dokken/kitchen_sandbox/a2abe14b87-lwrps-centos-67
       Creating verifier sandbox at /Users/rmoriz/.dokken/verifier_sandbox/a2abe14b87-lwrps-centos-67
       Building work image..
       Finished creating <lwrps-centos-67> (0m48.15s).
-----> Converging <lwrps-centos-67>...
       Creating kitchen sandbox in /Users/rmoriz/.dokken/kitchen_sandbox/a2abe14b87-lwrps-centos-67
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 5.6.4...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
[2017-04-24T21:10:02+00:00] WARN: *****************************************
[2017-04-24T21:10:02+00:00] WARN: Did not find config file: /opt/kitchen/client.rb, using command line options.
[2017-04-24T21:10:02+00:00] WARN: *****************************************
[2017-04-24T21:10:02+00:00] WARN: No cookbooks directory found at or above current directory.  Assuming /.
[2017-04-24T21:10:02+00:00] FATAL: Cannot load configuration from /opt/kitchen/dna.json
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Converge failed on instance <lwrps-centos-67>.  Please see .kitchen/logs/lwrps-centos-67.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration

kitchen setup lwrps-centos-67  4,01s user 1,64s system 10% cpu 55,907 total
```

**Comment by rmoriz** [comment]
This code path never gets executed…

https://github.com/someara/kitchen-dokken/blob/master/lib/kitchen/transport/dokken.rb#L78-L131

**Comment by someara** [comment]
I'll try and think of a way to set up this condition in tests.

**Comment by rmoriz** [comment]
I tried to setup Virtualbox/Vagrant in Travis but this didn't work because the VM does not provide hardware virtualization. :(

**Comment by rmoriz** [comment]
I've added a Docker-in-Docker instance which should expose the issue.

**Comment by someara** [comment]
The tests already run docker-in-docker.

https://github.com/someara/kitchen-dokken/blob/7fcb710e89fcc7a6b1643da4fe8e0d4506fa35e8/test/cookbooks/dokken_test/recipes/default.rb#L1-L4

... which works because we mount /var/lib/docker as a volume so it shows up as a block device and

https://github.com/someara/kitchen-dokken/blob/7fcb710e89fcc7a6b1643da4fe8e0d4506fa35e8/.kitchen.yml#L6

The code is then called explicitly with a tcp `DOCKER_HOST` here

https://github.com/someara/kitchen-dokken/blob/7fcb710e89fcc7a6b1643da4fe8e0d4506fa35e8/test/cookbooks/dokken_test/recipes/default.rb#L57-L65

**Comment by rmoriz** [comment]
But your setup is still a local docker instance, despite the use of tcp instead of a unix socket.

**Comment by someara** [comment]
@rmoriz can you check if this is fixed up for you in the latest version?
-s

**Comment by rmoriz** [comment]
@someara rebased master in the test PR #101 - looks like it's still failing

**Comment by jjasghar** [comment]
Running Docker in Docker, then running `dokken` inside that container to run kitchen seems (cough jenkins cough) seems to recreate this issue. If yall need another tester/verifier i'm here to help.

**Comment by rmoriz** [comment]
@jjasghar actually that is was PR #101 does

**Comment by someara** [comment]
I think I finally fixed this with in 2.4.1. 
Does it work for y'all?

**Comment by tas50** [comment]
Closing this out since we have the fix, but #101 will add a bit of testing, which we certainly want
